be/src/storage/vectorized/merge_iterator.cpp

```diff
@@ -226,16 +248,130 @@ ChunkIteratorPtr new_merge_iterator(const std::vector<ChunkIteratorPtr>& childre
     const static size_t kMaxChildrenSize = std::numeric_limits<uint16_t>::max();
 
     if (children.size() <= kMaxChildrenSize) {
-        return std::make_shared<HeapMergeIterator>(children);
+        return std::make_shared<HeapMergeIterator>(children, source_masks);
     }
     std::vector<ChunkIteratorPtr> sub_merge_iterators;
     sub_merge_iterators.reserve((children.size() + kMaxChildrenSize - 1) / kMaxChildrenSize);
     for (size_t i = 0; i < children.size(); i += kMaxChildrenSize) {
         size_t j = std::min(i + kMaxChildrenSize, children.size());
         std::vector<ChunkIteratorPtr> v(children.begin() + i, children.begin() + j);
-        sub_merge_iterators.emplace_back(new_merge_iterator(v));
+        sub_merge_iterators.emplace_back(new_heap_merge_iterator(v));
+    }
+    return new_heap_merge_iterator(sub_merge_iterators);
+}
+
+// Merge iterator based on source masks.
+// The order of rows is determinate by mask sequence.
+class MaskMergeIterator final : public MergeIterator {
+public:
+    explicit MaskMergeIterator(std::vector<ChunkIteratorPtr> children, RowSourceMaskBuffer* mask_buffer,
+                               std::vector<RowSourceMask>* source_masks)
+            : MergeIterator(std::move(children)),
+              _chunks(_children.size()),
+              _mask_buffer(mask_buffer),
+              _source_masks(source_masks) {
+#ifndef NDEBUG
```

Comment:
This macro condition is unneeded, `DCHECK` will be compiled away in release building mode.